Use the userid obtained from checkPassword after login is validated

sharidas opened the pull request:
The user id sent to the after event should use the id obtained
from checkPassword method, which validates the login credentials
passed to loginWithPassword method. This change would help the
the subscriber of the after event to get the appropriate userid.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
<!--- Describe your changes in detail -->
Use the uid returned from `checkPassword()` instead of the uid passed to `loginWithPassword()`. Updated the after event data accordingly to pick the uid.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
https://github.com/owncloud/core/issues/31509

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Use the uid returned from `checkPassword()` instead of the uid passed to `loginWithPassword()`. Updated the after event data accordingly to pick the uid.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- [x] With the unit test written for `loginWithPassword()`

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

butonic commented on lib/private/User/Session.php:
```diff
@@ -487,7 +487,7 @@ public function tryBasicAuthLogin(IRequest $request) {
 	 */
 	private function loginWithPassword($uid, $password) {
-		return $this->emittingCall(function () use (&$uid, &$password) {
+		return $this->emittingCall(function (&$afterArray) use (&$uid, &$password) {
```
try `return $this->emittingCall(function (&$afterArray) use ($uid, $password) {`

butonic commented on lib/private/User/Session.php:
```diff
@@ -502,6 +502,7 @@ private function loginWithPassword($uid, $password) {
 				$this->manager->emit('\OC\User', 'postLogin', [$user, $password]);
 				if ($this->isLoggedIn()) {
 					$this->prepareUserLogin($firstTimeLogin);
+					$afterArray['uid'] = $user->getUID();
```
why not pass in the user object? `$afterArray['user'] = $user;` the old hook does the same thing in https://github.com/owncloud/core/pull/31515/files#diff-9c47cee6ac987e5256aeee509f91ddb1R502

An object is much more useful than the uid as string

butonic commented on tests/lib/User/SessionTest.php:
```diff
@@ -1201,4 +1201,53 @@ public function testTryAuthModuleLogin($expectedReturn, array $modules) {
+
+		$this->assertEquals('user.beforelogin', $calledBeforeEvent[0]);
+		$this->assertInstanceOf(GenericEvent::class, $calledBeforeEvent[1]);
+		$this->assertArrayHasKey('uid', $calledBeforeEvent[1]);
```
in the before event this is not the uid, but the login (email or username). please rename the key to `login`. `uid` is just plain wrong. changeit in https://github.com/owncloud/core/pull/31515/files#diff-9c47cee6ac987e5256aeee509f91ddb1R517

butonic commented on lib/private/User/Session.php:
```diff
@@ -487,7 +487,7 @@ public function tryBasicAuthLogin(IRequest $request) {
 	 * @throws LoginException if an app canceld the login process or the user is not enabled
 	 */
 	private function loginWithPassword($uid, $password) {
```
so ... $uid here is actually the login (username or email). rename it to `$login` (also update the phpdoc)

butonic reviewed (changes requested):
thx for trying to fix this. uid vs login is not clear everywhere. see individual comments

PVince81 commented:
@sharidas comments adressed ? can you finish this ?

PVince81 commented:
@sharidas please also leave the old keys in there for compatibility in case apps use this already.
Mark as deprecated in the code for now and say so in the symfony events documentation (or have the doc only mention the new keys)

- [ ] update symfony doc ticket

PVince81 commented:
@butonic please recheck as you had concerns

butonic commented on lib/private/User/Session.php:
```diff
@@ -481,27 +481,34 @@ public function tryBasicAuthLogin(IRequest $request) {
 				if ($this->isLoggedIn()) {
 					$this->prepareUserLogin($firstTimeLogin);
+					$afterArray['user'] = $user;
+					$afterArray['uid'] = $user;
```
`$afterArray['uid'] = $user->getUID();` uid is the string ...

butonic commented on lib/private/User/Session.php:
```diff
@@ -513,7 +520,7 @@ private function loginWithPassword($uid, $password) {
 				throw new LoginException($message);
 			}
-		}, ['before' => ['uid' => $uid, 'password' => $password], 'after' => ['uid' => $uid, 'password' => $password]], 'user', 'login');
+		}, ['before' => ['login' => $login, 'uid' => $login, 'password' => $password], 'after' => ['user' => $login, 'uid' => $login, 'password' => $password]], 'user', 'login');
```
Just to indicate what needs to go where:
```php
['before' => [
  'login' => $login,
  'uid' => $login, // actually I'd love to deprecate this ... before login we don't know the uid. just here for compatability
  '_uid' => 'deprecated: please use \'login\', the real uid is not yet known' // maybe we can expose the deprecation this way?
  'password' => $password
], 'after' => [
  'login' => $login,
  'uid' => $user->getUid(), // uid = the string
  'user' => $user, // user = the object
  'password' => $password
]]
```

@PVince81 @DeepDiver1975 how do we deprecate uid in the preLogin hook params? We could also use an object that implements `toString()` which logs a warning if someone accesses it?

sharidas committed:
Use the userid obtained from checkPassword after login is validated

This change includes:
a) Change the function argument from 'uid' to 'login'. The php
doc is also updated.
b) before event argument would use 'login' to have loginname
c) after event argument would use 'user' instead of 'uid', because
'user' is an instance of IUser. We pass User object.

The user id sent to the after event should use the id obtained
from checkPassword method, which validates the login credentials
passed to loginWithPassword method. This change would help the
the subscriber of the after event to get the appropriate userid.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

jvillafanez commented on lib/private/User/Session.php:
```diff
@@ -486,28 +486,35 @@ public function tryBasicAuthLogin(IRequest $request) {
-			$this->manager->emit('\OC\User', 'preLogin', [$uid, $password]);
-			$user = $this->manager->checkPassword($uid, $password);
+	private function loginWithPassword($login, $password) {
+		return $this->emittingCall(function (&$afterArray) use ($login, $password) {
```
Could we simplify this and make it more standard? This is really confusing as it is.
This should as be as simple as:
```
send before event
do your stuff
send after event
```
You can easily read the function from top to bottom, as well as include extra parameters to the "after event" if needed.

Right now what we have is a big callback function and some parameters that it isn't clear where they're going to go. You have to go to the bottom of the function to know what parameters are going to be sent, and then scroll up to read the real function.
In addition, the `&$afterArray` make things worse: using references inside callbacks is quite dangerous because normally you don't use them, and also the code above must be prepared for it. If 90% of the calls don't use references but a 10% do, the first thing you think is that there might be a bug. Note that just checking the changes I thought it was dead code; I had to check the code of the "emittingCall" function to understand.

My point is that the function is structured in a counter-intuitive way. To understand a function you should just read through it, but this isn't happening with the current code.

sharidas replied:
Updated as per request. Unfortunately haven't updated the event name to camel case. The reason being I suspect there are listeners to the event in other app, like admin_audit.

butonic commented on lib/private/User/Session.php:
```diff
@@ -486,40 +486,47 @@ public function tryBasicAuthLogin(IRequest $request) {
+		$beforeEvent = new GenericEvent(null, ['login' => $login, 'uid' => $login, '_uid' => 'deprecated: please use \'login\', the real uid is not yet known', 'password' => $password]);
+		$this->eventDispatcher->dispatch('user.beforelogin', $beforeEvent);
+		$this->manager->emit('\OC\User', 'preLogin', [$login, $password]);
+		$user = $this->manager->checkPassword($login, $password);
```
add a newline before this line to separate event throwing code from the actual password check

sharidas replied:
done

butonic commented on lib/private/User/Session.php:
```diff
@@ -486,40 +486,47 @@ public function tryBasicAuthLogin(IRequest $request) {
+			$this->manager->emit('\OC\User', 'postLogin', [$user, $password]);
+			if ($this->isLoggedIn()) {
+				$this->prepareUserLogin($firstTimeLogin);
+				$afterEvent = new GenericEvent(null, ['user' => $user, 'uid' => $user->getUID(), 'password' => $password]);
```
add a newline before this line

sharidas replied:
done

butonic commented on lib/private/User/Session.php:
```diff
@@ -486,40 +486,47 @@ public function tryBasicAuthLogin(IRequest $request) {
+				$this->prepareUserLogin($firstTimeLogin);
+				$afterEvent = new GenericEvent(null, ['user' => $user, 'uid' => $user->getUID(), 'password' => $password]);
+				$this->eventDispatcher->dispatch('user.afterlogin', $afterEvent);
+				return true;
```
add a newline before this line

sharidas replied:
done

butonic reviewed (commented):
I like the increase in readability, now add some newlines I I am happy

phil-davis commented:
Backport ``stable10`` #31605
